# **Random Forest Regressor**

In [1]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
import numpy as np

In [3]:
import pandas as pd


In [4]:
url = 'https://raw.githubusercontent.com/SiddardhaShayini/Cyber-Crime-Analysis-and-Prediction-in-India/refs/heads/main/datasets/processed_dataset.csv'
df = pd.read_csv(url)

In [11]:
# Feature set including yearly data and engineered features
year_cols = [str(year) for year in range(2002, 2022)]
feature_cols = year_cols + ["State/UT_encoded", "Mean_2002_2021", "Std_2002_2021", "YoY_2021", "Total-Scale"]


In [12]:
# X and y
X = df[feature_cols]
y = df["Total"]

In [13]:
# Split dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [14]:
# Model
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)


RandomForestRegressor(random_state=42)

In [15]:
# Predict
y_pred = model.predict(X_test)

In [16]:
# Evaluation
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

In [17]:
print("Random Forest with All Important Features:")
print(f"MAE: {mae:.2f}")
print(f"RMSE: {rmse:.2f}")
print(f"R² Score: {r2:.2f}")

Random Forest with All Important Features:
MAE: 12235.75
RMSE: 27442.65
R² Score: 0.88


In [18]:
from sklearn.metrics import mean_absolute_percentage_error

mape = mean_absolute_percentage_error(y_test, y_pred)
print(f"Mean Absolute Percentage Error (MAPE): {mape * 100:.2f}%")


Mean Absolute Percentage Error (MAPE): 26.88%


### Performance Metrics:

| **Metric**                      | **Value**       |
|----------------------------------|------------------|
| Mean Absolute Error (MAE)        | 12,235.75        |
| Root Mean Squared Error (RMSE)   | 27,442.65        |
| R² Score (Coefficient of Determination) | 0.88             |
| Mean Absolute Percentage Error (MAPE) | **26.88%**       |



# **XGBoost Regressor**

In [19]:
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score


In [20]:
# Features and target
features = [
    "State/UT_encoded", "Total", "Total-Scale",
    "Mean_2002_2021", "Std_2002_2021", "YoY_2021"
] + [str(year) for year in range(2002, 2022)]
target = "2021"

In [21]:
X = df[features]
y = df[target]

In [22]:
# Split data
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [23]:
# Initialize and train model
xgb_model = xgb.XGBRegressor(n_estimators=100, learning_rate=0.1, max_depth=5, random_state=42)
xgb_model.fit(X_train, y_train)


XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=0.1, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=5, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=100, n_jobs=None,
             num_parallel_tree=None, random_state=42, ...)

In [24]:
# Predictions
y_pred = xgb_model.predict(X_test)

In [26]:
# MAE
mae = mean_absolute_error(y_test, y_pred)

# RMSE
rmse = mean_squared_error(y_test, y_pred) ** 0.5

# R²
r2 = r2_score(y_test, y_pred)

# MAPE
mape = (abs((y_test - y_pred) / y_test).mean()) * 100

In [27]:
# Print Results
print("Model: XGBoost Regressor")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")
print(f"Mean Absolute Percentage Error (MAPE): {mape:.2f}%")

Model: XGBoost Regressor
Mean Absolute Error (MAE): 5520.25
Root Mean Squared Error (RMSE): 14918.25
R² Score: 0.23
Mean Absolute Percentage Error (MAPE): 357.49%


### XGBoost Regressor Evaluation Report

| Metric                        | Value       |
|------------------------------|-------------|
| **MAE**                      | 5520.25     |
| **RMSE**                     | 14918.25    |
| **R² Score**                 | 0.23        |
| **MAPE**                     | 357.49%     |



# **Gradient Boosting Regressor**

In [28]:
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score


In [29]:
# Initialize and train the model
gbr_model = GradientBoostingRegressor(random_state=42)
gbr_model.fit(X_train, y_train)

GradientBoostingRegressor(random_state=42)

In [30]:
# Make predictions
y_pred = gbr_model.predict(X_test)


In [32]:
# Evaluation
mae = mean_absolute_error(y_test, y_pred)
rmse = mean_squared_error(y_test, y_pred) ** 0.5
r2 = r2_score(y_test, y_pred)
mape = (abs((y_test - y_pred) / y_test).mean()) * 100

In [33]:
print(f"Model: Gradient Boosting Regressor")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")
print(f"Mean Absolute Percentage Error (MAPE): {mape:.2f}%")

Model: Gradient Boosting Regressor
Mean Absolute Error (MAE): 226.05
Root Mean Squared Error (RMSE): 299.33
R² Score: 1.00
Mean Absolute Percentage Error (MAPE): 388.54%


### Gradient Boosting Regressor

| Metric                        | Value       |
|------------------------------|-------------|
| **MAE**                      | 226.05   |
| **RMSE**                     | 299.33    |
| **R² Score**                 | 1.00        |
| **MAPE**                     | 388.54%     |



# **LightGBM**

In [35]:
import lightgbm as lgb
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

In [36]:
# Define features and target
features = ['2002', '2003', '2004', '2005', '2006', '2007', '2008', '2009',
            '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017',
            '2018', '2019', '2020', 'State/UT_encoded',
            'Total', 'Mean_2002_2021', 'Std_2002_2021', 'YoY_2021']


In [37]:
target = '2021'

In [38]:
# Split data
from sklearn.model_selection import train_test_split
X = df[features]
y = df[target]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [39]:
# Train LightGBM Regressor
model = lgb.LGBMRegressor(random_state=42)
model.fit(X_train, y_train)

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 31, number of used features: 0
[LightGBM] [Info] Start training from score 3046.709677
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the s

LGBMRegressor(random_state=42)

In [40]:
# Predict
y_pred = model.predict(X_test)


In [41]:
# Evaluation
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100

In [42]:
# Print Results
print("Model: LightGBM Regressor")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")
print(f"Mean Absolute Percentage Error (MAPE): {mape:.2f}%")

Model: LightGBM Regressor
Mean Absolute Error (MAE): 8778.85
Root Mean Squared Error (RMSE): 17718.61
R² Score: -0.09
Mean Absolute Percentage Error (MAPE): 10351.35%


### LightGBM Regressor

| Metric                        | Value       |
|------------------------------|-------------|
| **MAE**                      | 8778.85   |
| **RMSE**                     | 17718.61   |
| **R² Score**                 | -0.09    |
| **MAPE**                     | 10351.35%     |



# **Dense Neural Networkwith Keras (TensorFlow)**

In [46]:
!pip install numpy==1.23.5 --quiet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 42.5 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
blosc2 3.2.1 requires numpy>=1.26, but you have numpy 1.23.5 which is incompatible.
jax 0.5.2 requires numpy>=1.25, but you have numpy 1.23.5 which is incompatible.
xarray 2025.1.2 requires numpy>=1.24, but you have numpy 1.23.5 which is incompatible.
albumentations 2.0.5 requires numpy>=1.24.4, but you have numpy 1.23.5 which is incompatible.
chex 0.1.89 requires numpy>=1.24.1, but you have numpy 1.23.5 which is incompatible.
scikit-image 0.25.2 requires numpy>=1.24, but you have numpy 1.23.5 which is incompatible.
albucore 0.0.23 requires numpy>=1.24.4, but you have numpy 1.23.5 which is incompatible.
bigframes 1.42.0 requires numpy>=1.24.0, but you have numpy 1.23.5 which is incompatible.
pymc 5.21.1 requires numpy>=1.25.0, but you h

In [47]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

OpenCV bindings requires "numpy" package.
Install it via command:
    pip install numpy


In [48]:
# Features and target
features = [
    'State/UT_encoded', 'Total', 'Total-Scale',
    'Mean_2002_2021', 'Std_2002_2021', 'YoY_2021'
]
target = '2021'

In [49]:
X = df[features]
y = df[target]


In [50]:
# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [51]:
# Feature scaling (important for neural networks)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [52]:
# Build the DNN model
model = Sequential([
    Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    Dense(32, activation='relu'),
    Dense(1)  # Output layer for regression
])

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [53]:
# Compile the model
model.compile(optimizer='adam', loss='mse', metrics=['mae'])


In [54]:
# Train the model
history = model.fit(X_train_scaled, y_train, epochs=100, batch_size=8, validation_split=0.1, verbose=0)


In [55]:
# Predictions
y_pred = model.predict(X_test_scaled).flatten()


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 244ms/step


In [56]:
# Evaluation
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)
mape = np.mean(np.abs((y_test - y_pred) / y_test)) * 100


In [57]:
# Results
print("Model: Dense Neural Network (Keras)")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")
print(f"Mean Absolute Percentage Error (MAPE): {mape:.2f}%")

Model: Dense Neural Network (Keras)
Mean Absolute Error (MAE): 7612.53
Root Mean Squared Error (RMSE): 18026.14
R² Score: -0.13
Mean Absolute Percentage Error (MAPE): 427.96%


### Dense Neural Network (Keras)

| Metric                        | Value       |
|------------------------------|-------------|
| **MAE**                      | 7612.53   |
| **RMSE**                     | 18026.14   |
| **R² Score**                 | -0.13    |
| **MAPE**                     | 427.96%     |



# **Combine Gradient Boosting + Random Forest**

In [58]:
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, VotingRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

In [59]:
# Define individual models
rf = RandomForestRegressor(n_estimators=100, random_state=42)
gb = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, random_state=42)


In [60]:
# Combine using VotingRegressor
ensemble = VotingRegressor([('rf', rf), ('gb', gb)])


In [61]:
# Train
ensemble.fit(X_train, y_train)

VotingRegressor(estimators=[('rf', RandomForestRegressor(random_state=42)),
                            ('gb', GradientBoostingRegressor(random_state=42))])

In [62]:
# Predict
y_pred = ensemble.predict(X_test)

In [63]:
# Evaluation
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)
mape = np.mean(np.abs((y_test - y_pred) / np.clip(np.abs(y_test), 1e-8, None))) * 100


In [64]:
# Results
print("Combined Model: Random Forest + Gradient Boosting")
print(f"Mean Absolute Error (MAE): {mae:.2f}")
print(f"Root Mean Squared Error (RMSE): {rmse:.2f}")
print(f"R² Score: {r2:.2f}")
print(f"Mean Absolute Percentage Error (MAPE): {mape:.2f}%")

Combined Model: Random Forest + Gradient Boosting
Mean Absolute Error (MAE): 1120.97
Root Mean Squared Error (RMSE): 2696.36
R² Score: 0.97
Mean Absolute Percentage Error (MAPE): 814.92%


### Combined Model: Random Forest + Gradient Boosting

| Metric                        | Value       |
|------------------------------|-------------|
| **MAE**                      | 1120.97   |
| **RMSE**                     | 2696.36   |
| **R² Score**                 | 0.97    |
| **MAPE**                     | 814.92%     |



# **Tune Gradient Boosting**

In [65]:
import numpy as np
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint, uniform

In [66]:
# Check for small y_test values
print("Min y_test value:", y_test.min())
print("Count of y_test values < 1000:", (y_test < 1000).sum())

Min y_test value: 5
Count of y_test values < 1000: 5


In [67]:
# SMAPE function
def smape(y_true, y_pred):
    return 100 / len(y_true) * np.sum(2 * np.abs(y_pred - y_true) / (np.abs(y_true) + np.abs(y_pred)))


In [68]:
# Hyperparameter tuning
param_dist = {
    "n_estimators": randint(100, 500),
    "learning_rate": uniform(0.01, 0.3),
    "max_depth": randint(3, 10),
    "subsample": uniform(0.6, 0.4)
}

In [71]:
gbr = GradientBoostingRegressor(random_state=42)
random_search = RandomizedSearchCV(
    gbr,
    param_distributions=param_dist,
    n_iter=50,
    cv=5,
    scoring='neg_mean_absolute_error',
    random_state=42,
    n_jobs=1
)


In [72]:
# Fit on training data
random_search.fit(X_train, y_train)

RandomizedSearchCV(cv=5, estimator=GradientBoostingRegressor(random_state=42),
                   n_iter=50, n_jobs=1,
                   param_distributions={'learning_rate': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7f206c341590>,
                                        'max_depth': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7f2069a94a50>,
                                        'n_estimators': <scipy.stats._distn_infrastructure.rv_discrete_frozen object at 0x7f2068064610>,
                                        'subsample': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x7f20636b5a50>},
                   random_state=42, scoring='neg_mean_absolute_error')

In [73]:
# Best model
best_gbr = random_search.best_estimator_

In [75]:
# Predict
y_pred_best_gbr = best_gbr.predict(X_test)


In [76]:
# Evaluation
mae = mean_absolute_error(y_test, y_pred_best_gbr)
rmse = np.sqrt(mean_squared_error(y_test, y_pred_best_gbr))
r2 = r2_score(y_test, y_pred_best_gbr)
smape_value = smape(y_test, y_pred_best_gbr)

In [77]:
# Report
print("Tuned Gradient Boosting Results:")
print(f"MAE: {mae:.2f}")
print(f"RMSE: {rmse:.2f}")
print(f"R² Score: {r2:.2f}")
print(f"SMAPE: {smape_value:.2f}%")

Tuned Gradient Boosting Results:
MAE: 261.74
RMSE: 347.33
R² Score: 1.00
SMAPE: 49.33%


### Summary of ML & DL Models:

| Model                   | MAE     | RMSE     | R² Score | MAPE     |
|------------------------|---------|----------|----------|----------|
| **Gradient Boosting**  | **226** | **299**  | **1.00** | 388.54%  |
| **Random Forest**      | 12235   | 27442    | 0.88     | 26.88%   |
| **XGBoost**            | 5520    | 14918    | 0.23     | 357.49%  |
| **LightGBM**           | 8778    | 17718    | -0.09    | 10351.35%|
| **Dense Neural Network with Keras (TensorFlow)**           | 7612.53    | 18026.14    | -0.13    | 427.96%|
| **Random Forest + Gradient Boosting**           | 1120.97    | 2696.36    | 0.97    | 814.92%|
| **Tuned Gradient Boosting**            | 261.74    | 347.33    | 1.00     | 49.33% (SMAPE) |

### Observations:
1. **Gradient Boosting**:
   - **MAE**: 226, **RMSE**: 299, **R²**: 1.00, **MAPE**: 388.54%.
   - *Pros*: Exceptionally low errors with perfect fit (**R² = 1.00**).  
   - *Cons*: Extremely high **MAPE** limits practical usage, likely due to poor predictions for small target values.

2. **Random Forest**:
   - **MAE**: 12235, **RMSE**: 27442, **R²**: 0.88, **MAPE**: 26.88%.  
   - *Pros*: Acceptable **MAPE**, solid R².  
   - *Cons*: Very high absolute errors (MAE, RMSE), making it less competitive.

3. **XGBoost**:
   - **MAE**: 5520, **RMSE**: 14918, **R²**: 0.23, **MAPE**: 357.49%.
   - *Cons*: Poor R² and high errors across metrics; underfitting may be an issue.

4. **LightGBM**:
   - **MAE**: 8778, **RMSE**: 17718, **R²**: -0.09, **MAPE**: 10351.35%.  
   - *Cons*: Negative **R²** suggests underperformance and weak predictions.

5. **Dense Neural Network (Keras/TensorFlow)**:
   - **MAE**: 7612.53, **RMSE**: 18026.14, **R²**: -0.13, **MAPE**: 427.96%.
   - *Cons*: Negative R² and high errors indicate architecture or training issues.

6. **Random Forest + Gradient Boosting**:
   - **MAE**: 1120.97, **RMSE**: 2696.36, **R²**: 0.97, **MAPE**: 814.92%.  
   - *Pros*: Excellent **R²**, low MAE/RMSE.  
   - *Cons*: High **MAPE**, similar to standalone Gradient Boosting.

7. **Tuned Gradient Boosting**:
   - **MAE**: 261.74, **RMSE**: 347.33, **R²**: 1.00, **SMAPE**: 49.33%.  
   - *Pros*: Stellar performance, perfect fit (R² = 1.00), significant improvement in relative error (MAPE → SMAPE).

---

**Best Performing Model**:
   - **Tuned Gradient Boosting** stands out with the lowest errors and balanced metrics. Its SMAPE of **49.33%** is reasonable and better than others.


### Save the Tuned Gradient Boosting Regressor Model

In [78]:
import joblib

# Save the tuned model
joblib.dump(best_gbr, 'tuned_gradient_boosting_model.pkl')


['tuned_gradient_boosting_model.pkl']